File: src/ee/core/internal/MessageBridge.cpp
//
//  BridgeManager.cpp
//  ee_core
//
//  Created by Zinge on 3/28/17.
//
//

#include <cassert>
#include <mutex>

#include "ee/core/internal/MessageBridge.hpp"

namespace ee {
namespace core {
MessageBridge& MessageBridge::getInstance() {
    static MessageBridge sharedInstance;
    return sharedInstance;
}

MessageBridge::MessageBridge() {
}

MessageBridge::~MessageBridge() {
}

std::string MessageBridge::call(const std::string& tag) {
    return call(tag, "");
}

std::string MessageBridge::callCpp(const std::string& tag,
                                   const std::string& msg) {
    std::lock_guard<SpinLock> guard(handlerLock_);
    auto iter = handlers_.find(tag);
    if (iter == handlers_.cend()) {
        assert(false);
        return "";
    }
    return iter->second(msg);
}

bool MessageBridge::registerHandler(const MessageHandler& handler,
                                    const std::string& tag) {
    std::lock_guard<SpinLock> guard(handlerLock_);
    if (handlers_.count(tag) > 0) {
        assert(false);
        return false;
    }
    handlers_[tag] = handler;
    return true;
}

bool MessageBridge::deregisterHandler(const std::string& tag) {
    std::lock_guard<SpinLock> guard(handlerLock_);
    if (handlers_.count(tag) == 0) {
        assert(false);
        return false;
    }
    handlers_.erase(tag);
    return true;
}
} // namespace core
} // namespace ee

File: proj.android_studio/ee-x-core/src/main/java/com/ee/core/MessageBridge.java
package com.ee.core;

import android.support.annotation.NonNull;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by Zinge on 3/29/17.
 */

public class MessageBridge {
    private static final Logger _logger = new Logger(MessageBridge.class.getName());

    /**
     * Registered handlers.
     */
    private Map<String, MessageHandler> _handlers;

    private final Object _handlerLock;

    /**
     * Thread-safe singleton pattern.
     */
    private static class Holder {
        private static final MessageBridge Instance = new MessageBridge();
    }

    /**
     * Gets a message bridge instance.
     *
     * @return A message bridge instance.
     */
    public static MessageBridge getInstance() {
        return Holder.Instance;
    }

    private MessageBridge() {
        _handlers = new HashMap<>();
        _handlerLock = new Object();
    }


    /**
     * Calls a handler from Java with a message.
     *
     * @param tag The unique tag of the handler.
     * @param msg The message.
     * @return Reply message from Java.
     */
    @NonNull
    private String call(@NonNull String tag, @NonNull String message) {
        MessageHandler handler = findHandler(tag);
        if (handler == null) {
            _logger.error("call: " + tag + " doesn't exist!");
            return "";
        }
        return handler.handle(message);
    }

    /**
     * Statically calls a handler from Java with a message.
     *
     * @param tag The unique tag of the handler.
     * @param msg The message.
     * @return Reply message from Java.
     */
    @NonNull
    private static String staticCall(@NonNull String tag, @NonNull String message) {
        return getInstance().call(tag, message);
    }

    /**
     * Calls a handler from C++ without a message.
     *
     * @param tag The unique tag of the handler.
     * @return Reply message from c++.
     */
    @NonNull
    public String callCpp(@NonNull String tag) {
        return callCpp(tag, "");
    }

    /**
     * Calls a handler from C++ with a message.
     *
     * @param tag The unique tag of the handler.
     * @param msg The message.
     * @return Reply message from C++.
     */
    @NonNull
    public native String callCpp(@NonNull String tag, @NonNull String message);

    /**
     * Registers a new handler to receive messages from C++.
     *
     * @param handler The handler.
     * @param tag     The unique tag of the handler.
     * @return Whether the registration was successful.
     */
    public boolean registerHandler(MessageHandler handler, @NonNull String tag) {
        synchronized (_handlerLock) {
            if (_handlers.containsKey(tag)) {
                _logger.error("registerHandler: " + tag + " already exists!");
                return false;
            }
            _handlers.put(tag, handler);
            return true;
        }
    }

    /**
     * Deregisters an existing handler not to receive messages from C++.
     *
     * @param tag The unique
     * @return Whether the deregistration was successful.
     */
    public boolean deregisterHandler(@NonNull String tag) {
        synchronized (_handlerLock) {
            if (!_handlers.containsKey(tag)) {
                _logger.error("deregisterHandler: " + tag + " doesn't exist!");
                return false;
            }
            _handlers.remove(tag);
            return true;
        }
    }

    private MessageHandler findHandler(@NonNull String tag) {
        synchronized (_handlerLock) {
            if (!_handlers.containsKey(tag)) {
                return null;
            }
            return _handlers.get(tag);
        }
    }
}
